### ISSUE TITLE
LimitNPROC line in /lib/systemd/system/openvpn@.service has to be commented out (maybe relevant to issue #199?)

### ISSUE THREAD

**Issue by frankblob** [Nyr/openvpn-install#206]
The LimitNPROC line in /lib/systemd/system/openvpn@.service has to be commented out in order to be able to start OpenVPN.

This issue manifests on a fresh install and update of Ubuntu 16.04.1, clean iptables and latest version of the installer script. Only OS post-install modification is 1) added sudo user, 2) changed to pub key login via ssh and 3) removed Apache (sudo service apache2 stop + sudo apt purge --auto-remove apache2*). Then 4) straight to downloading and running the OpenVPN script.

Relevant log files:
Oct  1 21:38:17 openvpnserver systemd[1]: Starting OpenVPN connection to server...
Oct  1 21:38:17 openvpnserver systemd[1]: Starting OpenVPN service...
Oct  1 21:38:17 openvpnserver systemd[1]: Started OpenVPN service.
Oct  1 21:38:17 openvpnserver ovpn-server[229]: OpenVPN 2.3.10 x86_64-pc-linux-gnu [SSL (OpenSSL)] [LZO] [EPOLL] [PKCS11] [MH] [IPv6] built on Feb  2 2016
Oct  1 21:38:17 openvpnserver systemd[1]: Failed to start OpenVPN connection to server.

I install OpenVPN across Ubuntu 14.04 and 16.04 via the script a lot of the time. OpenVPN stopped running after OS updates and system reboots (a while ago, maybe two months). I couldn't get OpenVPN started again immediately and I some OpenVPN backup servers on routers, so I didn't get around to troubleshooting and fixing it until now.

**Comment by Nyr** [comment]
Thanks a lot for the report.

This issue could be related to both #199 and #200. I see it has been previously [reported](http://askubuntu.com/questions/747023/systemd-fails-to-start-openvpn-in-lxd-managed-16-04-container) by other OpenVPN users in Ubuntu.

Sadly, I am unable to reproduce. It is pretty clear that this isn't related to the script, but if it is widespread, I could attempt a fix during setup. The biggest problem is that systemd service files can be overwritten during upgrades and I shouldn't be the one fixing them to start with ;)

Let's see if there are more users affected by this issue. For now, at least we have a workaround.

**Comment by stefancrain** [comment]
Wow, @Nyr I had a comment ready to submit with exactly that info but I couldn't recreate with my Ubuntu VM @ work. I'll see if I can recreate tonight.

**Comment by Nyr** [comment]
Thanks @stefancrain. Even if we are able to reproduce this issue, the right person to contact would likely be the Ubuntu packages maintainer, assuming that this isn't also happening for Debian.

**Comment by maxisme** [comment]
+1 moved to a clean install on ubuntu 14.04 and worked fine.

**Comment by Nyr** [comment]
Can't do much as long as we can't reproduce, but I'm pretty sure that this is a problem with the systemd script itself.

If that's the case = out of my hand.

**Comment by frankblob** [comment]
I had this issue manifesting on a fresh Ubuntu 16.04.1 running on a Windows Azure dedicated VPS and on CrownCloud, as well as on a fresh Ubuntu 14.04.5 and 16.04.1 on **some** LowEndSpirit NAT servers, but not all of them. For LES servers it appears to be location dependent, as LES Sweden worked (until the location was dropped), while Dallas didn't.

So it seems well out of your hands. Not related to the script. So, it's merely a work-around option for those who need it.

**Comment by maxisme** [comment]
@Nyr sorry I know it is not part of your script :) just wanting this 'Issue' to be kept up to date for if someone finds a solution. Really strange. Also have you heard of [streisand](https://github.com/jlund/streisand) I believe I had the same issue with them!

**Comment by maxisme** [comment]
@frankblob are LowEndSpirit any good? decided to go with i-83.net who are pretty solid. Also tried hostsailor who are awful.

**Comment by frankblob** [comment]
@maxisme Let's see if @Nyr enjoys derailing issues into hosting reviews...but yes, I have only good things to say about LowEndSpirit. It's stated pretty straight forward and upfront what you're getting into, so forum support reponse times should be something you accept and enjoy before signing up. Usual precautions to sharing an IP with numerous others should be considered, as with others hosting companies, of course. With that said, I really enjoy using LES servers for the purposes I have, as well as the ideology/culture behind the initiative.

**Comment by Nyr** [comment]
I would highly appreciate if one of the affected users can report the issue [here](https://bugs.launchpad.net/ubuntu/+source/openvpn). That should be the right way to get it fixed.

PS: no problem with going a little off-topic as long as the issue tracker doesn't get constantly spammed. LowEndSpirit is a nice project, as long as you understand and accept the limitations.

**Comment by maxisme** [comment]
[here](https://bugs.launchpad.net/ubuntu/+source/openvpn/+bug/1631104) is the bug report

**Comment by Nyr** [comment]
I would like to get confirmation from one of you guys able to reproduce: is this issue happening in Ubuntu but NOT in Debian?

The Ubuntu bug tracker has been updated but they are claiming that the issue is upstream (well, the change was), even though no Debian user has reported this AFAIK.

**Comment by Nyr** [comment]
Just a quick update to tell that I am also unable to reproduce on Azure with Ubuntu 16.04 as @frankblob reported. So sadly I still have zero clue as to why could OpenVPN spawn more than 10 processes.

Affected users could try to edit `/etc/openvpn/server.conf` and remove the `user` and `group` lines and see if that helps.

**Comment by Nactive** [comment]
@Nyr I tried with the most minimal configuration file I know and I still have the problem

> dev tun
> ifconfig 10.8.0.1 10.8.0.2
> secret static.key
> log-append  openvpn.log

Relevant log file information
> Tue Oct 25 21:21:20 2016 OpenVPN 2.3.10 x86_64-pc-linux-gnu [SSL (OpenSSL)] [LZO] [EPOLL] [PKCS11] [MH] [IPv6] built on Feb  2 2016
> Tue Oct 25 21:21:20 2016 library versions: OpenSSL 1.0.2g  1 Mar 2016, LZO 2.08
> Tue Oct 25 21:21:20 2016 daemon() failed or unsupported: Resource temporarily unavailable (errno=11)
> Tue Oct 25 21:21:20 2016 Exiting due to fatal error

In my case I am running in a VPS on a OpenVZ machine though so that might have something todo with it.

**Comment by Nyr** [comment]
@Nactive is this Ubuntu or Debian? If Ubuntu, could you try to reproduce the issue in Debian?

OpenVZ is fine.

**Comment by Nactive** [comment]
I am running Ubuntu 16.04 with kernel 2.6.32-042stab112.15. I don't have a debian machine to try to reproduce it.

If I increase LimitNPROC to 20 openvpn starts without problems. On 15 I still have errors.

**Comment by f1nnix** [comment]
OpenVZ @ Host1Plus (16.04, 2.6 kernel) seems to be also affected. Commenting `LimitNPROC=15` or increasing value as @Nactive said fixed the bug.

**Comment by Nyr** [comment]
@f1nnix do you know if it happens also in Debian or not? With that same environment, I mean.

**Comment by f1nnix** [comment]
Hi, @Nyr , finally had a chance to check on Debian with OpenVZ:

    # cat /etc/debian_version
    8.6

    # uname -a
    Linux 2.6.32-042stab120.11 #1 SMP Wed Nov 16 12:05:45 MSK 2016 x86_64 GNU/Linux

    # free -h
                 total       used       free     shared    buffers     cached
    Mem:          256M       242M        13M       6.6M         0B       212M
    -/+ buffers/cache:        29M       226M
    Swap:         128M       6.8M       121M

    # ps ax|grep openvpn
    5755 ?        Ss     0:00 /usr/sbin/openvpn --daemon ovpn-server --status /run/openvpn/server.status 10 --cd /etc/openvpn --config /etc/openvpn/server.conf
    5775 pts/0    S+     0:00 grep openvpn

Everthing works as expected.

**Comment by Nyr** [comment]
@f1nnix just to confirm: is this happening in the same container where Ubuntu was having issues?

Thanks a lot.

**Comment by f1nnix** [comment]
Yep. I've just reinstalled OS in the same container.

IP, hoster, RAM, storage and, obviously, kernel are the same:

* **hoster:** host1plus(dot)com
* **instance type:** Amber
* **CPU:** 0.5 Core
* **RAM:** 256 MB
* **HDD:** 20 GB
* **Bandwidth:** 500 GB

**Comment by Nyr** [comment]
I appreciate that, thanks again.

**Comment by pooley999** [comment]
Just an update. Im having the same problems with Ubuntu 14 and 16. Script runs fine and everything looks ok but no response from server. All usual rules checked, Port forwarding all ok, iptables configured properly. No dice unfortuantely.

Clean build updated and just intrpduced to your script

**Comment by Nyr** [comment]
@pooley999 if yours is really the same issue (service not starting up), there is nothing I can do, it has already been reported to those who should try to fix it.

If you just can't connect and don't understand why, that's a different issue. If so, please let's keep this thread on topic.

**Comment by schafdog** [comment]
Just been hit by this issue on debian running on OpenVZ host.

The service has been working fine for years with the LimitNPROC set. 

Had the service restarting every minute due to another configuration suddenly configured on multiple servers. After fixing this problem I was hit by the NPROC limit. I think a reboot would fix it, but haven't tried yet. Is there anything I can try before rebooting

**Comment by Nyr** [comment]
I an closing this since ultimately this is not an issue with the script and there isn't much more which can be done from here.

I understand that this is still an issue, but it's not with my project. I can't force the package managers to fix their systemd service file.

**Comment by Nyr** [comment]
Package maintainers had been aware of this issue for a [long time](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=819919) and they don't seem to care much, since it only happens with OpenVZ (and likely other container based virtual environments).

Maybe we could try opening a new bug on their tracker? The issue affects plenty of people and maybe they could reconsider the situation...

**Comment by frankblob** [comment]
Just had it happen to a 16.04.3 w/ 3GB RAM VPS (OpenVZ, yes) outside LES - specifically at CrownCloud in Frankfurt. The VPS had been running for almost two years with OpenVPN, but the OpenVPN server stopped working yesterday. Sure enough, commenting out the LimitPROC and rebooting fixed it immediately. It seems the issue is may be spreading with some recent upstream updates and hitting (OpenVZ) servers that were - for some reason - unaffected until now.

**Comment by Nyr** [comment]
@frankblob I suspect that it could have something to do with the fact that OpenVZ isn't a fully isolated environment and LimitNPROC is somehow taking in account processes from other containers, but I can't say for sure since I don't have a clue of how LimitNPROC works.

Fixing the issue is in the hands of package managers, the script could of course edit the service file but it's not a good idea nor a proper solution.

**Comment by Nyr** [comment]
Good explanation of the issue by someone who understands how LimitNPROC works:

https://github.com/systemd/systemd/issues/6011#issuecomment-304617744

**Comment by Axuk** [comment]
Is there now a fix to this issue? Been looking around for the past couple of hours and cannot seem to find a fix.

**Comment by Nyr** [comment]
@Axuk 

Just do `systemctl edit openvpn-server@server.service` and put something like this:
```
[Service]
LimitNPROC=100
```

I am not sure but maybe just leaving the `LimitNPROC` line with a blank value instead of a higher one would clear it, you could try that too.

**Comment by wwwindisch** [comment]
It might be out of topic, but perhaps helpful:

I tried to run OpenVPN on a fresh Ubuntu Server 18.04 (VM @ Hetzner) and faced exactly the same symptoms as described above. But playing around with LimitNPROC didn't help.  After some trial and error, I figured out  that actually the PrivateTmp setting is causing the error on Ubuntu (Debian is fine). Hence, my solution is:
`systemctl edit openvpn-server@server.service`

and then
```
[Service]
PrivateTmp=false
```

I have no idea, why PrivateTmp is causing this problem and I understand that just disabling it could possibly cause security issues...so be careful.
